proof/clear_cofactor_p2.saw

```diff
@@ -0,0 +1,190 @@
+  do {
+    unfolding ["clear_cofactor_bls12381_g2", "clear_cofactor_E2_impl"];
+    // Affinify goal
+    simplify (addsimps (concat [psi_psi_double_equiv_thm, times_minus_z_thm, psi_equiv_thm'] e2_curve_op_thms) (cryptol_ss()));
+
+    // Selectively remove polymorphic curve parameters from goal
+    simplify (addsimps [affinify_E'_thm, e2_scalar_mult_thm] empty_ss);
+
+    // Algebraic goal manipulation
+    simplify (addsimps [sub_to_add_neg_p2_thm, psi_neg_thm, neg_neg_p2_thm, mult_commutes_p2_thm, pull_mult_neg_p2_thm, push_neg_add_p2_thm] empty_ss);
+
+    // Sort adds
+    simplify (addsimps (concat [add_E'_thm, add_normalize_p2_thm] add_sort_p2_thms) empty_ss);
+
+    w4_unint_z3 ["affinify_E'", "E'", "e2_scalar_mult"];
+  };
+
+
+// Prove POINTonE2_add_n_dbl is equivalent to add_n_dbl.  Need to prove each
+// `n` separately because `n` is a type parameter on the add_n_dbl Cryptol
+// spec.
+let POINTonE2_add_n_dbl_spec n = do {
+  (out, out_ptr) <- ptr_to_fresh "out" POINTonE2_type;
+  (p, p_ptr) <- ptr_to_fresh_readonly "p" POINTonE2_type;
+  crucible_precond {{ POINTonE2_invariant out }};
+  crucible_precond {{ POINTonE2_invariant p }};
+  crucible_execute_func [ out_ptr, p_ptr, (crucible_term {{ `n : [64] }}) ];
+
+  let out' = {{ POINTonE2_rep (add_n_dbl`{n} (POINTonE2_abs out) (POINTonE2_abs p)) }};
+
+  crucible_postcond {{ POINTonE2_invariant out' }};
+
+  crucible_points_to out_ptr (crucible_term out');
+};
+POINTonE2_add_n_dbl_ovs <- for [2, 3, 9, 32, 16]
+  (\x -> custom_verify "POINTonE2_add_n_dbl"
+    [POINTonE2_double_alias_1_2_ov, POINTonE2_dadd_null_alias_1_2_ov]
+    (POINTonE2_add_n_dbl_spec x)
+    do {
+      unfolding ["POINTonE2_abs", "POINTonE2_rep", "POINTonE2_invariant"];
+      simplify fp2_simpset;
+      w4_unint_z3 fp2_unints;
+    });
+
+// Prove POINTonE2_times_minus_z matches the times_minus_z spec
+let POINTonE2_times_minus_z_spec = do {
+  out_ptr <- crucible_alloc POINTonE2_type;
+  (in_, in_ptr) <- ptr_to_fresh_readonly "in" POINTonE2_type;
+  crucible_precond {{ POINTonE2_invariant in_ }};
+  crucible_execute_func [out_ptr, in_ptr];
+  let out = {{ POINTonE2_rep (times_minus_z (POINTonE2_abs in_)) }};
+  crucible_postcond {{ POINTonE2_invariant out }};
+  crucible_points_to out_ptr (crucible_term out);
+};
+POINTonE2_times_minus_z_ov <- custom_verify "POINTonE2_times_minus_z"
+  (concat [POINTonE2_double_ov] POINTonE2_add_n_dbl_ovs)
+  POINTonE2_times_minus_z_spec
+  do {
+    unfolding ["POINTonE2_abs", "POINTonE2_rep", "POINTonE2_invariant"];
+    simplify fp2_simpset;
+    w4_unint_z3 fp2_unints;
+  };
+
+let clear_cofactor_spec = do {
+  out_ptr <- crucible_alloc POINTonE2_type;
+  (inp, inp_ptr) <- ptr_to_fresh_readonly "in" POINTonE2_type;
+  crucible_precond {{ POINTonE2_invariant inp }};
+
+  crucible_execute_func [out_ptr, inp_ptr];
+
+  crucible_points_to out_ptr (crucible_term {{ POINTonE2_rep (clear_cofactor_E2_impl (POINTonE2_abs inp)) }});
```

Comment:
I usually prefer to put the more abstract function in the spec, here that would be ` clear_cofactor_bls12381_g2 ` I think.  The proof of the override is no harder; you just start by using the `equiv` theorem to replace the abstract function with the more implementation-oriented one, and you've got that theorem already.

Reply:
That can only be done with `llvm_postcond` though. Is that what you mean?

Btw, we should switch to the `llvm_` prefix (`crucible_` is deprecated).

Reply:
Oops, yes you are right.  I hadn't looked at the equiv theorem statement closely enough.  So this has to stay as it is.  That's OK and we've done it this way elsewhere.